engine/crates/partial-caching/src/output/shapes/mod.rs

```diff
@@ -154,12 +151,61 @@ impl<'a> ConcreteShape<'a> {
 }
 
 #[derive(Clone, Copy)]
-#[allow(dead_code)] // Will be using this after GB-6949
 pub struct PolymorphicShape<'a> {
     shapes: &'a OutputShapes,
     id: ObjectShapeId,
 }
 
+impl<'a> PolymorphicShape<'a> {
+    pub(crate) fn concrete_shape_for_typename(
+        &self,
+        typename: &str,
+        type_relationships: &dyn TypeRelationships,
+    ) -> ConcreteShape<'a> {
+        let ObjectShapeRecord::Polymorphic {
+            type_conditions,
+            fallback,
+        } = &self.shapes.objects[self.id.0 as usize]
+        else {
+            unreachable!()
+        };
+
+        fn search(
+            typename: &str,
+            conditions: &[TypeConditionId],
+            fallback: ConcreteShapeId,
+            shapes: &OutputShapes,
+            type_relationships: &dyn TypeRelationships,
+        ) -> ConcreteShapeId {
+            let condition_match = conditions.iter().find(|id| {
+                type_relationships
+                    .type_condition_matches(shapes.type_conditions[id.0 as usize].type_condition.as_str(), typename)
+            });
+
+            match condition_match {
+                Some(id) => {
+                    let node = &shapes.type_conditions[id.0 as usize];
+                    search(
```

Comment:
A bit weird to have a search within a search when you already did a `find()` imho